### PR TITLE
Comparison with previous

### PR DESCRIPTION
**Pull request by omittones**
@cbovar Here is the project demonstrating differences between old CPU version and new CPU version. It has two training sets, XOR and Circle Region.
I will continue debugging, maybe I can figure out whats the issue. When you have time, please verify that the test is sound, maybe I missed something.

Circe Region Set:
circled-region-set

E.g. Circe Region outputs these losses of old (A) vs new (B) implementations:

Epoch   
000       A == 0.062 vs B == 0.406 ----> DIFF  0.344
001       A == 0.068 vs B == 0.375 ----> DIFF  0.308
002       A == 0.066 vs B == 0.367 ----> DIFF  0.301
003       A == 0.058 vs B == 0.367 ----> DIFF  0.309
004       A == 0.049 vs B == 0.369 ----> DIFF  0.320
005       A == 0.042 vs B == 0.372 ----> DIFF  0.331
006       A == 0.035 vs B == 0.376 ----> DIFF  0.342
007       A == 0.029 vs B == 0.382 ----> DIFF  0.353
008       A == 0.025 vs B == 0.389 ----> DIFF  0.365
009       A == 0.021 vs B == 0.399 ----> DIFF  0.378
010       A == 0.018 vs B == 0.412 ----> DIFF  0.393
011       A == 0.016 vs B == 0.425 ----> DIFF  0.410
012       A == 0.014 vs B == 0.443 ----> DIFF  0.430
013       A == 0.012 vs B == 0.466 ----> DIFF  0.454
014       A == 0.010 vs B == 0.494 ----> DIFF  0.484
015       A == 0.009 vs B == 0.525 ----> DIFF  0.516
016       A == 0.007 vs B == 0.567 ----> DIFF  0.560
017       A == 0.006 vs B == 0.624 ----> DIFF  0.618
018       A == 0.005 vs B == 0.697 ----> DIFF  0.691
019       A == 0.005 vs B == 0.800 ----> DIFF  0.795
020       A == 0.004 vs B == 0.928 ----> DIFF  0.924
021       A == 0.004 vs B == 1.056 ----> DIFF  1.052
022       A == 0.003 vs B == 1.173 ----> DIFF  1.170
023       A == 0.003 vs B == 1.254 ----> DIFF  1.251
024       A == 0.002 vs B == 1.289 ----> DIFF  1.287
025       A == 0.002 vs B == 1.286 ----> DIFF  1.284
026       A == 0.002 vs B == 1.259 ----> DIFF  1.258
027       A == 0.002 vs B == 1.216 ----> DIFF  1.215
028       A == 0.001 vs B == 1.164 ----> DIFF  1.163
029       A == 0.001 vs B == 1.109 ----> DIFF  1.108
030       A == 0.001 vs B == 1.037 ----> DIFF  1.035
031       A == 0.001 vs B == 0.947 ----> DIFF  0.946
032       A == 0.001 vs B == 0.836 ----> DIFF  0.835
033       A == 0.001 vs B == 0.695 ----> DIFF  0.695
034       A == 0.001 vs B == 0.530 ----> DIFF  0.529
035       A == 0.001 vs B == 0.426 ----> DIFF  0.425
036       A == 0.001 vs B == 0.344 ----> DIFF  0.343
037       A == 0.001 vs B == 0.261 ----> DIFF  0.261
038       A == 0.001 vs B == 0.229 ----> DIFF  0.228
039       A == 0.000 vs B == 0.205 ----> DIFF  0.204
040       A == 0.000 vs B == 0.199 ----> DIFF  0.198
041       A == 0.000 vs B == 0.195 ----> DIFF  0.195
042       A == 0.000 vs B == 0.191 ----> DIFF  0.190
043       A == 0.000 vs B == 0.189 ----> DIFF  0.189
044       A == 0.000 vs B == 0.187 ----> DIFF  0.187
045       A == 0.000 vs B == 0.182 ----> DIFF  0.182
046       A == 0.000 vs B == 0.175 ----> DIFF  0.175
047       A == 0.000 vs B == 0.168 ----> DIFF  0.168
048       A == 0.000 vs B == 0.162 ----> DIFF  0.162
049       A == 0.000 vs B == 0.155 ----> DIFF  0.155

### CODE REVIEW

**Comment by cbovar** [comment]
By changing changing [here](https://github.com/cbovar/ConvNetSharp/blob/master/src/ConvNetSharp.Volume/Double/Volume.cs#L357)
`this.Storage.Map((input, outputG) => outputG * input, outputGradient.Storage, inputGradient.Storage);` into 
`this.Storage.Map((input, outputG) => outputG, outputGradient.Storage, inputGradient.Storage);` 
and using a batch size of 1, I get the same loss for previous and current version.
I may have made a mistake in the softmax gradient computation.

For batch size > 1, I believe the weights and gradients are identical for both versions but the loss is computed differently (average of loss over the batch for current version)

**Comment by cbovar** [comment]
Current version Softmax gradient computation gives the same results as Cudnn which is reassuring. The  computation was probably wrong in the previous version but this is also suprising because it would mean that [ConvNetJS computation](https://github.com/karpathy/convnetjs/blob/master/src/convnet_layers_loss.js#L65) is wrong

**Comment by omittones** [comment]
@cbovar Yeah, possibly ConvNetJS has a slight bug. The test that checks gradients using Gradient Checker breaks if I replace `outputG * input` to `outputG`.

As for loss reported, if I understood you correctly the old version didn't report average loss, just the loss of the last sample, but this one does, beacuse of [this line](https://github.com/cbovar/ConvNetSharp/blob/master/src/ConvNetSharp.Core/Training/TrainerBase.cs#L31l), so I added manual loss averaging for both trainers, making the report more consistent.

However, I've pushed settings variation that causes new impl. to break. From some reason, the FullConn layer outputs during training explode to Infinity. It happens consistently for XOR: BATCH_SIZE = 1, LEARN_RATE = 0.05, MOMENTUM = 0. Interesting thing is that loss keeps decreasing normally, but at one point the model just breaks :)

**Comment by cbovar** [comment]
@omittones 
I have made some modifications in commit 7fdef02 to have the same result from previous and latest.
Modifications includes:
- same random seed (123)
- same way of filling up volume with random data
- same softmax gradient computation (buggy one)
- same bias pref bug (it was always 0 in previous version)

My guess is that when using the correct softmax gradient computation, the learning rate used in that experiment becomes too big, resulting in gradient explosion. 

Something annoying is that when using smaller learning rate, the loss in the latest version seems to converge to a higher value.

**Comment by cbovar** [comment]
@omittones I have commited a new implementation of SoftMaxGradient in [comparison-with-previous](https://github.com/cbovar/ConvNetSharp/tree/comparison-with-previous) branch. It looks more stable to me. when you have time, could you play with it and tell me what you think.

> 000       A == 0.666 vs B == 0.683 ----> DIFF  0.017
> 001       A == 0.596 vs B == 0.629 ----> DIFF  0.033
> 002       A == 0.447 vs B == 0.599 ----> DIFF  0.152
> 003       A == 0.350 vs B == 0.584 ----> DIFF  0.234
> 004       A == 0.309 vs B == 0.568 ----> DIFF  0.258
> 005       A == 0.251 vs B == 0.557 ----> DIFF  0.306
> 006       A == 0.233 vs B == 0.406 ----> DIFF  0.173
> 007       A == 0.221 vs B == 0.323 ----> DIFF  0.103
> 008       A == 0.213 vs B == 0.293 ----> DIFF  0.080
> 009       A == 0.208 vs B == 0.274 ----> DIFF  0.066
> 010       A == 0.196 vs B == 0.255 ----> DIFF  0.059
> 011       A == 0.178 vs B == 0.234 ----> DIFF  0.056
> 012       A == 0.188 vs B == 0.234 ----> DIFF  0.046
> 013       A == 0.170 vs B == 0.224 ----> DIFF  0.054
> 014       A == 0.166 vs B == 0.200 ----> DIFF  0.034
> 015       A == 0.155 vs B == 0.218 ----> DIFF  0.063
> 016       A == 0.172 vs B == 0.189 ----> DIFF  0.018
> 017       A == 0.152 vs B == 0.207 ----> DIFF  0.055
> 018       A == 0.149 vs B == 0.207 ----> DIFF  0.058
> 019       A == 0.175 vs B == 0.182 ----> DIFF  0.006
> 020       A == 0.159 vs B == 0.175 ----> DIFF  0.016
> 021       A == 0.159 vs B == 0.171 ----> DIFF  0.013
> 022       A == 0.139 vs B == 0.162 ----> DIFF  0.023
> 023       A == 0.139 vs B == 0.157 ----> DIFF  0.018
> 024       A == 0.136 vs B == 0.161 ----> DIFF  0.025
> 025       A == 0.157 vs B == 0.162 ----> DIFF  0.004
> 026       A == 0.132 vs B == 0.158 ----> DIFF  0.026
> 027       A == 0.125 vs B == 0.143 ----> DIFF  0.017
> 028       A == 0.156 vs B == 0.142 ----> DIFF  0.013
> 029       A == 0.139 vs B == 0.160 ----> DIFF  0.021
> 030       A == 0.143 vs B == 0.157 ----> DIFF  0.014
> 031       A == 0.146 vs B == 0.134 ----> DIFF  0.011
> 032       A == 0.136 vs B == 0.132 ----> DIFF  0.004
> 033       A == 0.136 vs B == 0.118 ----> DIFF  0.017
> 034       A == 0.124 vs B == 0.137 ----> DIFF  0.013
> 035       A == 0.133 vs B == 0.144 ----> DIFF  0.011
> 036       A == 0.118 vs B == 0.139 ----> DIFF  0.021
> 037       A == 0.154 vs B == 0.135 ----> DIFF  0.019
> 038       A == 0.108 vs B == 0.141 ----> DIFF  0.033
> 039       A == 0.114 vs B == 0.148 ----> DIFF  0.034
> 040       A == 0.106 vs B == 0.135 ----> DIFF  0.029
> 041       A == 0.122 vs B == 0.105 ----> DIFF  0.017
> 042       A == 0.114 vs B == 0.126 ----> DIFF  0.012
> 043       A == 0.130 vs B == 0.124 ----> DIFF  0.007
> 044       A == 0.126 vs B == 0.117 ----> DIFF  0.009
> 045       A == 0.111 vs B == 0.142 ----> DIFF  0.032
> 046       A == 0.139 vs B == 0.134 ----> DIFF  0.005
> 047       A == 0.123 vs B == 0.109 ----> DIFF  0.014
> 048       A == 0.115 vs B == 0.121 ----> DIFF  0.005
> 049       A == 0.121 vs B == 0.116 ----> DIFF  0.005

**Comment by omittones** [comment]
Looks much more stable now. I tried with a range of hyperparams, and was not able to force it to explode. It does however seem to converge more slowly for the same set of parameters, but I don't think it's a problem. Here are some of the runs that I executed:

```
SIZE:11-5
BATCH_SIZE:100
LEARN_RATE:0.1
MOMENTUM:0.2
-----------------------------------------------------------------------------------------------------
000       A == 0.744 vs B == 0.795 ----> DIFF  0.051
050       A == 0.515 vs B == 0.358 ----> DIFF  0.157
100       A == 0.491 vs B == 0.315 ----> DIFF  0.175
150       A == 0.227 vs B == 0.277 ----> DIFF  0.050
200       A == 0.083 vs B == 0.226 ----> DIFF  0.143
250       A == 0.058 vs B == 0.173 ----> DIFF  0.114
300       A == 0.046 vs B == 0.131 ----> DIFF  0.085
350       A == 0.038 vs B == 0.105 ----> DIFF  0.067
400       A == 0.033 vs B == 0.087 ----> DIFF  0.054
450       A == 0.029 vs B == 0.074 ----> DIFF  0.045

SIZE:5-5
BATCH_SIZE:10
LEARN_RATE:0.2
MOMENTUM:0
-----------------------------------------------------------------------------------------------------
000       A == 0.692 vs B == 0.648 ----> DIFF  0.045
050       A == 0.100 vs B == 0.071 ----> DIFF  0.028
100       A == 0.062 vs B == 0.056 ----> DIFF  0.006
150       A == 0.039 vs B == 0.052 ----> DIFF  0.013
200       A == 0.039 vs B == 0.037 ----> DIFF  0.002
250       A == 0.062 vs B == 0.028 ----> DIFF  0.034
300       A == 0.013 vs B == 0.047 ----> DIFF  0.035
350       A == 0.011 vs B == 0.092 ----> DIFF  0.082
400       A == 0.008 vs B == 0.082 ----> DIFF  0.073
450       A == 0.010 vs B == 0.041 ----> DIFF  0.031

SIZE:10-9-8-7
BATCH_SIZE:30
LEARN_RATE:0.1
MOMENTUM:0
-----------------------------------------------------------------------------------------------------
000       A == 0.691 vs B == 1.198 ----> DIFF  0.507
010       A == 0.350 vs B == 0.289 ----> DIFF  0.060
020       A == 0.118 vs B == 0.140 ----> DIFF  0.022
030       A == 0.135 vs B == 0.103 ----> DIFF  0.033
040       A == 0.099 vs B == 0.083 ----> DIFF  0.017
050       A == 0.090 vs B == 0.068 ----> DIFF  0.022
060       A == 0.064 vs B == 0.071 ----> DIFF  0.006
070       A == 0.061 vs B == 0.064 ----> DIFF  0.003
080       A == 0.078 vs B == 0.056 ----> DIFF  0.022
090       A == 0.093 vs B == 0.051 ----> DIFF  0.042
100       A == 0.058 vs B == 0.043 ----> DIFF  0.015
110       A == 0.066 vs B == 0.043 ----> DIFF  0.023
120       A == 0.058 vs B == 0.051 ----> DIFF  0.007
130       A == 0.053 vs B == 0.035 ----> DIFF  0.017
140       A == 0.056 vs B == 0.044 ----> DIFF  0.011

SIZE:50-50
BATCH_SIZE:500
LEARN_RATE:0.1
MOMENTUM:0.4
-----------------------------------------------------------------------------------------------------
000       A == 0.680 vs B == 0.699 ----> DIFF  0.019
010       A == 0.509 vs B == 0.559 ----> DIFF  0.049
020       A == 0.446 vs B == 0.514 ----> DIFF  0.068
030       A == 0.405 vs B == 0.482 ----> DIFF  0.076
040       A == 0.379 vs B == 0.456 ----> DIFF  0.077
050       A == 0.362 vs B == 0.435 ----> DIFF  0.074
060       A == 0.349 vs B == 0.419 ----> DIFF  0.070
070       A == 0.339 vs B == 0.405 ----> DIFF  0.066
080       A == 0.329 vs B == 0.394 ----> DIFF  0.065
090       A == 0.318 vs B == 0.385 ----> DIFF  0.067
100       A == 0.306 vs B == 0.377 ----> DIFF  0.071
110       A == 0.294 vs B == 0.370 ----> DIFF  0.077
120       A == 0.279 vs B == 0.364 ----> DIFF  0.086
130       A == 0.260 vs B == 0.359 ----> DIFF  0.100
140       A == 0.241 vs B == 0.354 ----> DIFF  0.113
150       A == 0.222 vs B == 0.350 ----> DIFF  0.128
```

**Comment by omittones** [comment]
Accidentally closed the PR :(
